### PR TITLE
Shim Layer to support multiple Spark versions 

### LINKED ISSUE #355: [FEA] Support Multiple Spark versions in the same jar
It would be nice if we can have a single spark rapids jar that supports multiple versions of Spark (spark 3.0, spark 3.1, etc), including different distributions based on the same Spark version (GCP, Databricks, AWS, etc)

Types of incompatibilities it would need ti handle:
- Simple change in function definition in Spark
- Classes take different parameters 
- Classes move packages - harder if we inherit from them
- Catalyst expressions/execs/etc removed or added

Proposal to handle:
Create a ShimLoader class that looks at the spark version and then loads the proper rapids subclass. This can have multiple different classes. For instance a SparkShim layer that is wrapper around simple function calls, we could have a loader for different classes that share a base class. For instance GpuBroadcastNestedLoopJoinExec could have a GpuBroadcastNestedLoopJoinExecBase class and then specific versions GpuBroadcastNestedLoopJoinExec30 and GpuBroadcastNestedLoopJoinExec31. 

Each Spark version then gets its own submodule where all the specific classes for Spark3.0, Spark 3.1, Spark 3.0 databricks etc. all live.  The Loader will look at the version and then load the corresponding class.

Some spark distributions make things a bit harder in that we can’t compile it at the same time as Spark 3.0 and spark 3.1.   So for those jars we will have to build them separately and ideally pull them as a dependency to package up and distribute a single jar.  The Apache spark 3.0 and spark 3.1 jars can be built in a single build.

To handle expressions and exec being added and removed we can have the GpuOverrides call down into the ShimLoader to get any version specific ones.  So for instance TimeSub in Spark 3.1 got changed to TimeAdd.  The GpuOverrides definition of exprs would leave it out but then call into the shim layer to get specific versions (++ ShimLoader.getSparkShims.getExecs).

### PR DESCRIPTION
**Pull request by tgravescs**
fixes #355

This adds a shim layer to support multiple Spark versions. This PR adds the framework and support for Apache Spark 3.0.0, Apache Spark 3.0.1 and Apache Spark 3.10. All the tests pass for Spark 3.0.0 and 3.0.1 but there are a few failures still for 3.1.0. We can finish resolving those after this is merged. The problem I keep running into is upmerged to spark 3.1 as well as the plugin itself keeps having conflicts and requires continuous retesting and fixing those. So if we can get this part reviewed and in, I think it will be easier to finish the remaining things and allow multiple people to look at it.

Note that to shim the ShuffleManger, it requires the user to specify the spark package version right now.

A lot of the changes are around Join. HashJoin had changes which requires a lot of things to be copied into the shim layer. There may be ways it improve this to share code but would like to look more at that in followup, if you have ideas let me know. If you diff those files there should be very little differences. The other issue there is BuildSide, BuildRight, BuildLeft all moved packages.

Spark 3.1.0 had api changes for other things - TimeSub, ScalaUDF, MapOutputTracker, ShuffleManager, FileSourceScan, First and Last

For the Shimlayer itself we use service loaders for each of the versions. There is a lightweight loader class that first determines if the loader applies to that Spark version, then it has a buildShim function to load the entire shim. This keeps us from loading a bunch of classes we don't really need.

I added a profile spark30tests and spark31tests that you can specify to run the tests on different versions. Once we have things working we can hook that up to CI along with running integration tests on both versions. slf4j in the tests had conflicts with the cudf versions, so for now I made them pull in the newer version. I want to update CUDF to have the same version as well.

Note that I tested the core sql-plugin code by building against both Spark 3.0 and Spark 3.1 to ensure I didn't miss anything there. Tests were run against both versions. 1 unit test failure on 3.1 and 3.1 has some integration test failures - some FullOuter join tests, log test, and TimeSub tests,

I filed a few followup issues to look at the remaining items for commonize join code, changing loaders to use priority, investigating if we want version match to be strict or not, and investigate shuffle having common base class.

### CODE REVIEW

**Comment by tgravescs** [comment]
build

**Comment by tgravescs** [comment]
Note I'm in progress of running a final pass over integration tests now.

File: dist/pom.xml [hunk 1]
```diff
@@ -94,6 +108,30 @@
...
+            <goals>
+              <goal>run</goal>
+            </goals>
+          </execution>
+        </executions>
+        <configuration>
+          <launchers>
+            <launcher>
+              <id>update_rapids_config</id>
+              <mainClass>com.nvidia.spark.rapids.RapidsConf</mainClass>
```
Comment (revans2) [inline dist/pom.xml:127]: Can we have a follow on issue to update RapidsConf so we can find a way to generate the docs with info for multiple versions?  This only works for the current version.

Comment (tgravescs) [reply dist/pom.xml:127]: https://github.com/NVIDIA/spark-rapids/issues/417

**Comment by tgravescs** [comment]
Ran integration tests on 3.0.0 and 3.0.1 and all pass, 3.1.0 has a few failures that I expected and have issue to investigate: https://github.com/NVIDIA/spark-rapids/issues/416

**Pull request merged**